include/wayland/mir/wayland/wayland_base.h

```diff
@@ -37,12 +41,62 @@ class Resource
     };
 
     Resource();
-    virtual ~Resource() = default;
+    ~Resource();
 
     Resource(Resource const&) = delete;
     Resource& operator=(Resource const&) = delete;
+
+    auto destroyed_flag() const -> std::shared_ptr<bool>;
+
+private:
+    std::shared_ptr<bool> mutable destroyed;
+};
+
+/// A weak handle to a Wayland resource
+/// May only be safely used from the Wayland thread
+template<typename T>
+class Handle
+{
+public:
+    Handle(T* resource)
+        : resource{resource},
+          destroyed_flag{resource->destroyed_flag()}
+    {
+    }
+
+    Handle(Handle<T> const&) = default;
+    auto operator=(Handle<T> const&) -> Handle<T>& = default;
+
+    auto operator==(Handle<T> const& other) const -> bool
+    {
+        return resource == other->resource;
+    }
+
+    operator bool() const
+    {
+        return !*destroyed_flag;
+    }
+
+    auto value() const -> T&
+    {
+        if (*destroyed_flag)
+        {
+            fatal_error("Attempted to access destroyed Wayland resource");
```

Comment:
Does this need to be a `fatal_error`? Throwing an exception should result in it being caught and the associated client disconnected with an implementation error.

I'm not blocking on this, just flagging the question.

Reply:
I was on the fence about which to make it. I think an exception might be better.